cmd/kubeadm/app/cmd/join.go

```diff
@@ -619,6 +619,12 @@ func fetchInitConfigurationFromJoinConfiguration(cfg *kubeadmapi.JoinConfigurati
 		return nil, nil, err
 	}
 
+	// Retrieve the final KubeConfig file with the cluster name discovered after fetching the cluster configuration
```

Comment:
I don't think we should do discovery twice.
Instead I think we should change the TLS kubeconfig in place or create a new kubeconfig file from TLS kubeconfig but with a different cluster name (if you need prior art let me know)

This will allow to clean up the Discovery.For chain of calls from the clusterName field